xchange-stream-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingAdapters.java

```diff
@@ -19,58 +14,93 @@
+            LimitOrder last = target.last();
+            target.remove(last);
+        }
+    }
 
-  public static OrderBook adaptOrderbookMessage(
-      OrderBook orderBook, Instrument instrument, ArrayNode arrayNode) {
-    Streams.stream(arrayNode.elements())
-        .filter(JsonNode::isObject)
-        .forEach(
-            currentNode -> {
-              for (String key : BID_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.BID, currentNode.get(key))
-                      .forEach(orderBook::update);
+    public static OrderBook adaptOrderbookMessage(int depth, AtomicBoolean awaitingSnapshot, TreeSet<LimitOrder> bids, TreeSet<LimitOrder> asks, Instrument instrument, ArrayNode arrayNode) {
+        final AtomicLong expectedChecksum = new AtomicLong(0);
+        final AtomicReference<Date> lastTime = new AtomicReference<>(Date.from(Instant.EPOCH));
+        arrayNode.elements().forEachRemaining(currentNode -> {
+            if (awaitingSnapshot.get()) {
+                if (currentNode.has(BID_SNAPSHOT) && currentNode.has(ASK_SNAPSHOT)) {
+                    LOG.debug("Received snapshot, clearing book");
+                    bids.clear();
+                    asks.clear();
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_SNAPSHOT, bids);
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_SNAPSHOT, asks);
+                    awaitingSnapshot.set(false);
                 }
-              }
-              for (String key : ASK_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.ASK, currentNode.get(key))
-                      .forEach(orderBook::update);
+            } else {
+                if (currentNode.has(BID_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_UPDATE, bids);
                 }
-              }
-            });
-    return new OrderBook(
-        orderBook.getTimeStamp(),
-        Lists.newArrayList(orderBook.getAsks()),
-        Lists.newArrayList(orderBook.getBids()),
-        true);
-  }
+                if (currentNode.has(ASK_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_UPDATE, asks);
+                }
+            }
+            if (!awaitingSnapshot.get() && currentNode.has(CHECKSUM)) {
+                expectedChecksum.set(currentNode.get(CHECKSUM).asLong());
+            }
+        });
+
+        OrderBook result = new OrderBook(lastTime.get(), Lists.newArrayList(asks), Lists.newArrayList(bids), true);
+        long localChecksum = createCrcChecksum(asks, bids);
+        if (expectedChecksum.get() > 0 && expectedChecksum.get() != localChecksum) {
+            LOG.warn("Checksum does not match, expected {} but local checksum is {}", expectedChecksum.get(), localChecksum);
```

Comment:
Is it possible to log information which order book has checksum mismatch? IMO would be helpful